File: crates/mithril-oxide-sys/cpp/Dialect/Transform/IR/TransformTypes.cpp
#include "TransformTypes.hpp"

#include <mlir/Dialect/Transform/IR/TransformTypes.h>
#include <mlir/IR/MLIRContext.h>
#include <mlir/IR/Types.h>
#include <rust/cxx.h>


namespace mithril_oxide_sys::transform {

const void *AnyOpType_get(MLIRContext *ctx)
{
    return AnyOpType::get(ctx).getAsOpaquePointer();
}

const void *OperationType_get(MLIRContext *ctx, rust::Str name)
{
    return OperationType::get(
        ctx,
        mlir::StringRef(name.data(), name.length())
    ).getAsOpaquePointer();
}

const void *ParamType_get(MLIRContext *ctx, const void *type)
{
    return ParamType::get(ctx, mlir::Type::getFromOpaquePointer(type)).getAsOpaquePointer();
}

} // namespace mithril_oxide_sys::transform

File: crates/mithril-oxide-sys/src/Dialect/Transform/IR/TransformTypes.rs
pub use self::ffi::DeviceAsyncTokenType_get;

#[cxx::bridge]
mod ffi {
    #[namespace = "mlir"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/Transform/IR/TransformTypes.hpp");

        type MLIRContext = crate::IR::MLIRContext::MLIRContext;
    }

    #[namespace = "mithril_oxide_sys::sparse_tensor"]
    unsafe extern "C++" {
        include!("mithril-oxide-sys/cpp/Dialect/Transform/IR/TransformTypes.hpp");

        type c_void = crate::IR::Value::ffi::c_void;

        pub unsafe fn AnyOpType_get(context: *mut MLIRContext) -> *const c_void;
        pub unsafe fn OperationType_get(context: *mut MLIRContext, name: &str) -> *const c_void;
        pub unsafe fn ParamType_get(
            context: *mut MLIRContext,
            type_: *const c_void,
        ) -> *const c_void;
    }
}
